ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+  are global without a clear way to reference them (would have to be `method(:my_activity)`)
+
+- It is possible to allow multiple activities per class, however we've decided against this because
+  there's an extra complexity to the invocation (extra argument representing the method) and it
+  complicates the use of class-level "decorators" (e.g. for specifying activity name or other
+  options)
+
+
+## Defining a workflow
+
+Similar to activities, workflows are also represented as classes.
+
+`sum_workflow.rb`
+
+```ruby
+require 'sum_activity'
+
+# Similar to activities the class name will be used as a workflow name unless overridden
+class SumWorkflow < Temporal::Workflow
+  # Optionally override the name
+  workflow_name 'sum-workflow'
+
+  def execute(a, b)
+    # - Positional arguments passed to activity (inline with other SDKs)
+    # - Keyword arguments are reserved for options
+    # - This call is blocking
+    workflow.execute_activity(SumActivity, a, b, start_to_close_timeout: 10)
+
+    # - Starting an activity by its name rather than a class
+    # - The result of this activity will be used as a workflow result (Ruby standard)
+    workflow.execute_activity('sum-activity', a, b, start_to_close_timeout: 10)
+  end
+end
+```
+
+#### Alternative approaches
+
+- For the same exact reasons it is not feasible to use non-namespaced methods for workflow
+
+- Similar to activities it should be possible to define multiple workflows per class, however this
+  will not be allowed to avoid a perception of a shared state between workflows and allow for
+  class-level "decorators"
+
+
+## Start a workflow
+
+```ruby
+require 'temporal/client'
+require 'sum_workflow'
+
+# Connections are eager and a lazy option will be introduced later
+connection = Temporal::Connection.connect('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+# workflow is referenced by the class
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+handle.result # => 3
+```
+
+#### Alternative approaches
+
+- Starting a foreign workflow from a string name
+
+```ruby
+client.start_workflow('sum-workflow', 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+- Starting a foreign workflow with the help of a stub
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  # Optionally override the name of the workflow
+  workflow_name 'my.workflow'
+end
+
+client.start_workflow(MyWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+
+## Start a worker
+
+```ruby
+require 'temporal/worker'
+require 'sum_workflow'
+require 'some_other_workflow'
+require 'sum_activity'
+require 'some_other_activity'
+
+connection = Temporal::Connection.connect('localhost:7233')
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [SumWorkflow, SomeOtherWorkflow],
+  activities: [SumActivity, SomeOtherActivity]
+)
```

Comment:
I think it's gonna be important to provide an `activity_runner` here which defaults to `ThreadPoolRunner` or something. I think it's really important to be able to let people make async activities in their favorite async frameworks. Arguably it could be a per-activity runner instead of a per-worker activity runner.

Being able to run thousands of workflows and thousands of activities on just a couple of threads is very nice for scale.

Reply:
I see where you're coming from but I would punt on this for now.

Reply:
I'll make sure there's a clear separation of the runner, which we can later expose as a configurable argument. But it's probably something that we don't need in the lightweight proposal :)